### ISSUE TITLE
Error ModalRoute on subscribe routeObserver

### ISSUE THREAD

**Issue by 1AlexFix1** [flutter/flutter#78447]
Hello, 
Faced the problem that the official example for RouteObserver doesn't work.
Copied the code from here https://api.flutter.dev/flutter/widgets/RouteObserver-class.html

Error:

![Снимок экрана 2021-03-17 в 21 17 46](https://user-images.githubusercontent.com/44409006/111517593-48a32f00-8766-11eb-8ab2-3f61f4f8c838.png)

<details>
[✓] Flutter (Channel stable, 2.0.1, on macOS 11.2.3 20D91 darwin-arm, locale ru)
    • Flutter version 2.0.1 at /Users/a.kotkov/flutter
    • Framework revision c5a4b4029c (13 days ago), 2021-03-04 09:47:48 -0800
    • Engine revision 40441def69
    • Dart version 2.12.0

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/a.kotkov/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] Connected device (2 available)
    • iPhone SE (2nd generation) (mobile) • 6BF925FC-1734-4DCC-94A3-546FE65A2285 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-4 (simulator)
    • Chrome (web)                        • chrome                               • web-javascript • Google Chrome 89.0.4389.90

• No issues found!

</details>

**Comment by darshankawar** [comment]
@1AlexFix1 
Verified the same on latest stable (2.0.2) and didn't notice the issue mentioned.  There were no compile errors:

<img width="857" alt="Screenshot 2021-03-18 at 1 54 36 PM" src="https://user-images.githubusercontent.com/67046386/111594858-91ec9080-87f1-11eb-921a-9b7ba123565b.png">

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.0.2, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.0.2 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 8962f6dc68 (6 days ago), 2021-03-11 13:22:20 -0800
    • Engine revision 5d8bf811b3
    • Dart version 2.12.1

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.54.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (3 available)
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                               •
      web-javascript • Google Chrome 89.0.4389.90

! Doctor found issues in 1 category.

```

</details>

<details>
<summary> code sample </summary>

```
// Register the RouteObserver as a navigation observer.
import 'package:flutter/material.dart';

final RouteObserver<PageRoute> routeObserver = RouteObserver<PageRoute>();
void main() {
  runApp(MaterialApp(
    home: Container(),
    navigatorObservers: [routeObserver],
  ));
}

class RouteAwareWidget extends StatefulWidget {
  State<RouteAwareWidget> createState() => RouteAwareWidgetState();
}

// Implement RouteAware in a widget's state and subscribe it to the RouteObserver.
class RouteAwareWidgetState extends State<RouteAwareWidget> with RouteAware {

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    routeObserver.subscribe(this, ModalRoute.of(context));
  }

  @override
  void dispose() {
    routeObserver.unsubscribe(this);
    super.dispose();
  }

  @override
  void didPush() {
    // Route was pushed onto navigator and is now topmost route.
  }

  @override
  void didPopNext() {
    // Covering route was popped off the navigator.
  }

  @override
  Widget build(BuildContext context) => Container();

}

```

</details>

**Comment by 1AlexFix1** [comment]
> @ 1AlexFix1 Проверил
> то же самое на последней стабильной версии (2.0.2) и не заметил упомянутой проблемы. Ошибок компиляции не было:
> 
> <img alt="Снимок экрана 2021-03-18 в 1 54 36 PM" width="857" src="https://user-images.githubusercontent.com/67046386/111594858-91ec9080-87f1-11eb-921a-9b7ba123565b.png">
> 
> трепещущий доктор -v
> ```
> [✓] Flutter (Channel stable, 2.0.2, on Mac OS X 10.15.4 19E2269 darwin-x64,
>     locale en-GB)
>     • Flutter version 2.0.2 at /Users/dhs/documents/Fluttersdk/flutter
>     • Framework revision 8962f6dc68 (6 days ago), 2021-03-11 13:22:20 -0800
>     • Engine revision 5d8bf811b3
>     • Dart version 2.12.1
> 
> [!] Xcode - develop for iOS and macOS
>     • Xcode at /Applications/Xcode.app/Contents/Developer
>     • Xcode 12.3, Build version 12C33
>     ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
>         CocoaPods is used to retrieve the iOS and macOS platform side's plugin
>         code that responds to your plugin usage on the Dart side.
>         Without CocoaPods, plugins will not work on iOS or macOS.
>         For more info, see https://flutter.dev/platform-plugins
>       To upgrade see
>       https://guides.cocoapods.org/using/getting-started.html#installation for
>       instructions.
> 
> [✓] Chrome - develop for the web
>     • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome
> 
> [✓] VS Code (version 1.54.1)
>     • VS Code at /Applications/Visual Studio Code.app/Contents
>     • Flutter extension version 3.18.1
> 
> [✓] Connected device (3 available)
>     • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729 • ios
>       • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
>     • macOS (desktop)            • macos                                •
>       darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
>     • Chrome (web)               • chrome                               •
>       web-javascript • Google Chrome 89.0.4389.90
> 
> ! Doctor found issues in 1 category.
> ```
> 
> образец кода
> ```
> // Register the RouteObserver as a navigation observer.
> import 'package:flutter/material.dart';
> 
> final RouteObserver<PageRoute> routeObserver = RouteObserver<PageRoute>();
> void main() {
>   runApp(MaterialApp(
>     home: Container(),
>     navigatorObservers: [routeObserver],
>   ));
> }
> 
> class RouteAwareWidget extends StatefulWidget {
>   State<RouteAwareWidget> createState() => RouteAwareWidgetState();
> }
> 
> // Implement RouteAware in a widget's state and subscribe it to the RouteObserver.
> class RouteAwareWidgetState extends State<RouteAwareWidget> with RouteAware {
> 
>   @override
>   void didChangeDependencies() {
>     super.didChangeDependencies();
>     routeObserver.subscribe(this, ModalRoute.of(context));
>   }
> 
>   @override
>   void dispose() {
>     routeObserver.unsubscribe(this);
>     super.dispose();
>   }
> 
>   @override
>   void didPush() {
>     // Route was pushed onto navigator and is now topmost route.
>   }
> 
>   @override
>   void didPopNext() {
>     // Covering route was popped off the navigator.
>   }
> 
>   @override
>   Widget build(BuildContext context) => Container();
> 
> }
> ```

digging in the project, I found that if you raise
dart to 2.12(Enabled null safe) then 
there will be an error, if you put 2.7(Disable null safe) then everything is ok

**Comment by darshankawar** [comment]
@1AlexFix1 
I am already on `Dart version 2.12.1` when tried replicating the issue. Can you try by upgrading to 2.0.2 (stable) ?

**Comment by 1AlexFix1** [comment]
> @1AlexFix1
> I am already on `Dart version 2.12.1` when tried replicating the issue. Can you try by upgrading to 2.0.2 (stable) ?

Yes, I saw you have a higher version and updated, but the result is the same. I created a clean project from scratch, but no, the result is the same, mb - Android studio making brains?
What's your version of the idea?
my
Android Studio 4.1.2
Build # AI-201.8743.12.41.7042882, built December 20, 2020

**Comment by darshankawar** [comment]
<img width="492" alt="Screenshot 2021-03-19 at 11 45 23 AM" src="https://user-images.githubusercontent.com/67046386/111739182-a0e24a00-88a8-11eb-8017-0f6a56567e59.png">

**Comment by 1AlexFix1** [comment]
> @ 1AlexFix1
> Я уже участвовал, `Dart version 2.12.1`когда пытался воспроизвести проблему. А можно попробовать обновиться до 2.0.2 (стабильная)?

I'll clarify again, just in case. Have you changed the version of dart in pubspec.yaml?
like this environment:
  sdk: "> = 2.12.1 <3.0.0"
Because even if you ignore the error of type incorrect ModalRoute.of (context) returns nullable, however routeObserverуsubscribe accepts non-null parameters, which should already show an error in the basic example.

**Comment by darshankawar** [comment]
@1AlexFix1 I see what you are saying.

<img width="862" alt="Screenshot 2021-03-22 at 1 13 26 PM" src="https://user-images.githubusercontent.com/67046386/111956135-6d095d80-8b10-11eb-90a3-c14742e10418.png">


<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 2.1.0-13.0.pre.194, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-GB)
    • Flutter version 2.1.0-13.0.pre.194 at
      /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision e58ee0fbc7 (2 hours ago), 2021-03-22 13:06:06 +0800
    • Engine revision e6db3a275e
    • Dart version 2.13.0 (build 2.13.0-149.0.dev)

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.54.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (3 available)
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                               •
      web-javascript • Google Chrome 89.0.4389.90

! Doctor found issues in 1 category.

```

</details>

**Comment by marcosmko** [comment]
Hello! I'm also experiencing this problem. Is there any solution?
Tested in both stable, beta and master channels.

## Logs

<details>
<summary>Logs</summary>

```
[✓] Flutter (Channel beta, 2.1.0-12.2.pre, on macOS 11.2.3 20D91 darwin-x64,
    locale en-BR)
    • Flutter version 2.1.0-12.2.pre at /Users/marcoskobuchi/flutter
    • Framework revision 5bedb7b1d5 (9 days ago), 2021-03-17 17:06:30 -0700
    • Engine revision 711ab3fda0
    • Dart version 2.13.0 (build 2.13.0-116.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/marcoskobuchi/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/marcoskobuchi/Library/Android/sdk
    • ANDROID_SDK_ROOT = /Users/marcoskobuchi/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

• No issues found!

```

</details>

**Comment by marcosmko** [comment]
Hello! I discovered it is just a documentation error.

Instead of:
```
final RouteObserver<PageRoute> routeObserver = RouteObserver<PageRoute>();
```

you should use:
```
final RouteObserver<Route> routeObserver = RouteObserver<Route>();
```

In the provided example, PageRoute is a subclass of `ModalRoute`, then you cannot assign a `ModalRoute` to a `PageRoute` type.

Also, when you subscribe to observer, you need to unwrap the result of `ModalRoute.of`:
```
routeObserver.subscribe(this, ModalRoute.of(context)!);
```

**Comment by marcosmko** [comment]
documentation fixed on #77606

**Comment by maheshmnj** [comment]
I just ran this [sample example](https://api.flutter.dev/flutter/widgets/RouteObserver-class.html) on the official docs and I can confirm this issue is not reproducible on the latest stable channel with dart 2.14 with null safety enabled.

Thanks, @marcosmko for the clarification.

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.2.3, on macOS 11.4 20F71 darwin-arm, locale en-IN)
    • Flutter version 2.2.3 at /Users/mahesh/Documents/flutter
    • Framework revision f4abaa0735 (8 days ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.1.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 57.0.5
    • Dart plugin version 211.7233

[✓] VS Code (version 1.58.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.24.0

[✓] Connected device (4 available)
    • sdk gphone arm64 (mobile)  • emulator-5554                        • android-arm64  • Android 11 (API 30) (emulator)
    • iPhone 12 Pro Max (mobile) • 6A434744-B348-4FB3-AB10-131292174456 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-4
      (simulator)
    • macOS (desktop)            • macos                                • darwin-arm64   • macOS 11.4 20F71 darwin-arm
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 91.0.4472.114

• No issues found!
```
</details>
